lib/plugin-base.js

```diff
@@ -0,0 +1,337 @@
+  } else {
+    if (this.artifact) {
+      config = this.loadNamed(rootDir, env, this.artifact);
+    }
+  }
+  // Register as context.configurations.<plugin-name>
+  return this.configure(context, config);
+};
+
+PluginBase.prototype.configure = function(context, config) {
+  config = config || {};
+  // Register as context.configurations.<plugin-name>
+  if (!context.configurations) {
+    context.configurations = {};
+  }
+  context.configurations[this.name] = config;
+  return config;
+};
+
+PluginBase.prototype.merge = function(target, config, keyPrefix) {
+  return this.mergeObjects(target, config, keyPrefix);
+};
+
+/**
+ * Load named configuration.
+ * @param {String} rootDir Directory where to look for files.
+ * @param {String} env Environment, usually `process.env.NODE_ENV`
+ * @param {String} name
+ * @returns {Object}
+ */
+PluginBase.prototype.loadNamed = function(rootDir, env, name) {
+  var files = this.findConfigFiles(rootDir, env, name);
+  debug('Looking in dir %s for %s configs', rootDir, this.name);
+  if (files.length) {
+    debug('found %s %s files: %j', env, name, files);
+    files.forEach(function(f) {
+      debug('  %s', f);
+    });
+  }
+  var configs = this.loadConfigFiles(files);
+  var merged = this.mergeConfigurations(configs);
+
+  debug('merged %s %s configuration %j', env, name, merged);
+
+  return merged;
+};
+
+/**
+ * Search `rootDir` for all files containing configuration for `name`.
+ * @param {String} rootDir Root directory
+ * @param {String} env Environment, usually `process.env.NODE_ENV`
+ * @param {String} name Name
+ * @param {Array.<String>} exts An array of extension names
+ * @returns {Array.<String>} Array of absolute file paths.
+ */
+PluginBase.prototype.findConfigFiles = function(rootDir, env, name, exts) {
+  var master = ifExists(name + '.json');
+  if (!master && (ifExistsWithAnyExt(name + '.local') ||
+    ifExistsWithAnyExt(name + '.' + env))) {
+    g.warn('WARNING: Main config file "%s{{.json}}" is missing', name);
+  }
+  if (!master) return [];
+
+  var candidates = [
+    master,
+    ifExistsWithAnyExt(name + '.common'),
+    ifExistsWithAnyExt(name + '.local'),
+    ifExistsWithAnyExt(name + '.' + env),
+  ];
+
+  return candidates.filter(function(c) {
+    return c !== undefined;
+  });
+
+  function ifExists(fileName) {
+    var filePath = path.resolve(rootDir, fileName);
+    return util.fileExistsSync(filePath) ? filePath : undefined;
+  }
+
+  function ifExistsWithAnyExt(fileName) {
+    var extensions = exts || ['js', 'json'];
+    var file;
+    for (var i = 0, n = extensions.length; i < n; i++) {
+      file = ifExists(fileName + '.' + extensions[i]);
+      if (file) {
+        return file;
+      }
+    }
+  }
+};
+
+/**
+ * Load configuration files into an array of objects.
+ * Attach non-enumerable `_filename` property to each object.
+ * @param {Array.<String>} files
+ * @returns {Array.<Object>}
+ */
+PluginBase.prototype.loadConfigFiles = function(files) {
```

Comment:
AFAICT, this is a "protected" method that should be used only by PluginBase and descendants. The convention is to prefix names of non-public methods with underscore, i.e.

```js
PluginBase.prototype._loadConfigFiles
```